docs/proposals/backend-support-appset.md

```diff
@@ -0,0 +1,226 @@
+* The ApplicationSetController looks at all the Applications owned by the ApplicationSet (via ownerref or annotation):
+    Verify that the user has permission to act on all of the Applications currently managed by the ApplicationSet
+* If the above precondition is met, proceed to the next step, otherwise fail.
+* The ApplicationSet is generated and rendered into a template
+  * Look at the generated Applications, and make sure the user has appropriate privileges
+    * All the same checks done by the Create workflow, described above, are done here (user can access repo, cluster, etc
+* Finally, on success, the API server applies (kubectl apply) the requested change to the ApplicationSet (and the Applications).
+
+#### **ApplicationSets and AppProjects:**
+
+An important design constraint in this area: ApplicationSets do not belong to projects. They generate Applications that are a part of projects, but they themselves are not part of a project.
+
+1. Why not just include a project field on ApplicationSet CR?
+
+   A single ApplicationSet has the power to produce Applications within multiple projects, so it does not necessarily make sense to include an ApplicationSet within a single project.
+
+2. Why not just include a projects field (array of strings) on ApplicationSets, to allow it to belong to multiple projects?
+
+   This is getting closer to ideal, but still limits the expressive power of the ApplicationSet: it requires a user to specify, up front, what projects they expect to generate applications for. Eg you must statically define an ApplicationSet's projects. This excludes the scenario where the projects that a particular ApplicationSet will generate Applications for is truly dynamic, eg coming from a configuration file in Git, and thus not known at creation.
+
+3. We still need a way to limit the scope of user actions against ApplicationSets. Even though ApplicationSets don't belong to a project, we still need to prevent users from modifying ApplicationSets that they don't have RBAC access to. By looking at the projects of child objects, we achieve the same goal. We don't want users to be able to delete ApplicationSets that they don't have RBAC access to. But, since ApplicationSets don't have a project, we need some way to tell what applications/projects an ApplicationSet manages. So we use the applications and projects within it.
+
+#### **Command Design:**
+
+```shell
+argocd appset create "(filename.yaml)"
+argocd appset delete "(applicationset resource name)"
+argocd appset apply  "(filename.yaml)"
+```
+
+This proposal assumes that the ApplicationSet controller is still an optional, standalone install. Thus all argocd appset commands should fail if the ApplicationSet controller is not installed. (The Argo CD API server would check if the ApplicationServer is running by looking for a deployment with a specific 'applicationset controller' annotation, or similar mechanism.)
```

Comment:
I think it's fine to naiively deploy the appset resource regardless of whether the controller is running.